### PR TITLE
Add feature_grad_mult argument to HuBERTPretrainModel

### PR DESCRIPTION
**Pull request by nateanl**
In Wav2Vec2 and HuBERT model training, the convolutional feature extraction layers use `group_norm` for normalization in `Base` model, while they use `layer_norm` in `Large` and `XLarge` models. For `Base` model, the gradients of feature extraction layers will be unstable in pre-training, thus we need to scale down the gradient by multiplying 0.1.

In this PR, we add such argument to `HuBERTPretrainModel` to control the gradient of feature extractor layers. We also put the argument in the factory functions (`hubert_pretrain_base`, `hubert_pretrain_large`, and `hubert_pretrain_xlarge`. The reason is in finetuning, the feature extractor's parameters are fixed, we can multiply the gradient with 0.0 to avoid back propagating gradients.

### CODE REVIEW

File: torchaudio/models/wav2vec2/model.py [hunk 1]
```diff
@@ -910,6 +915,11 @@ def hubert_pretrain_model(
 
             This option corresponds to ``final_dim`` from ``fairseq``.
 
+        feature_grad_mult (float):
+            Whether to scale down the gradient of convolutional feature extraction layers.
```
Comment (hwangjeff) [inline torchaudio/models/wav2vec2/model.py:919]: the language here implies that `feature_grad_mult` is of type `bool` when it's actually of type `float`. something like "factor to scale convolutional feature extraction layer gradients by" would be clearer

File: torchaudio/models/wav2vec2/model.py [hunk 2]
```diff
@@ -144,11 +144,13 @@ def __init__(
         wav2vec2: Wav2Vec2Model,
         mask_generator: Module,
         logit_generator: Module,
+        feature_grad_mult: float,
```
[the remark this answers is not in the document]
Comment (hwangjeff) [reply torchaudio/models/wav2vec2/model.py:147]: should add to the docstring above

File: torchaudio/models/wav2vec2/model.py (hunk 1 above)
Comment (mthrok) [reply torchaudio/models/wav2vec2/model.py:919]: Also probably worth mentioning that forward path is not affected.

File: torchaudio/models/wav2vec2/model.py [hunk 3]
```diff
@@ -712,6 +723,7 @@ def hubert_pretrain_model(
     skip_nomask: bool,
     num_classes: int,
     final_dim: int,
+    feature_grad_mult: float,
```
Comment (mthrok) [inline torchaudio/models/wav2vec2/model.py:726]: Should this be `float` or `Optional[float]`?

File: torchaudio/models/wav2vec2/model.py [hunk 4]
```diff
@@ -137,18 +137,27 @@ class HuBERTPretrainModel(Module):
...
         logit_generator: Module,
+        feature_grad_mult: Optional[float],
     ):
         super().__init__()
         self.wav2vec2 = wav2vec2
         self.mask_generator = mask_generator
         self.logit_generator = logit_generator
+        assert (
+            0.0 <= feature_grad_mult <= 1.0
+        ), f"The value of `feature_grad_mult` must be between [0, 1]. Found {feature_grad_mult}"
```
Comment (mthrok) [inline torchaudio/models/wav2vec2/model.py:159]: I think it should be exclusive in both end.
It does not make much sense if it is 0 or 1.

File: torchaudio/models/wav2vec2/model.py [hunk 5]
```diff
@@ -1033,6 +1060,7 @@ def hubert_pretrain_large(
     encoder_ff_interm_dropout: float = 0.0,
     encoder_dropout: float = 0.0,
     encoder_layer_drop: float = 0.0,
+    feature_grad_mult=1.0,
```
Comment (mthrok) [inline torchaudio/models/wav2vec2/model.py:1063]: If it's 1.0, gradient is not modified so why not default to `None`?

File: torchaudio/models/wav2vec2/model.py (hunk 4 above)
Comment (nateanl) [reply torchaudio/models/wav2vec2/model.py:159]: It makes sense when it is 0. In fine-tuning case, the parameters of feature extractor are fixed, which can be done by setting it to 0

File: torchaudio/models/wav2vec2/model.py (hunk 5 above)
Comment (nateanl) [reply torchaudio/models/wav2vec2/model.py:1063]: Sure, will change it.

File: torchaudio/models/wav2vec2/model.py (hunk 4 above)
Comment (mthrok) [reply torchaudio/models/wav2vec2/model.py]: I think in such cases, freezing the entire model using PyTorch's API is more appropriate. That way, the backward path is stopped. Relying on `feature_grad_mult =0` means that backward path is still computed.

Comment (nateanl) [reply torchaudio/models/wav2vec2/model.py]: There are two stages in the fine-tuning, in the first stage, only `aux` layer has gradient, the rest of parameters are fixed. In the second stage, the feature extractor's parameters are fixed. The train loop looks like:
```python
if self.global_step <= self.fix_encoder_updates:
    with torch.no_grad():
        x, out_len = self.model.wav2vec2.feature_extractor(waveforms, audio_lengths)
        padding_mask = components._get_padding_mask(x, out_len)
        x, attention_mask = self.model.wav2vec2.encoder._preprocess(x, out_len)
        x, mask = self.model.mask_generator(x, padding_mask)
        x = self.model.wav2vec2.encoder.transformer(x, attention_mask=attention_mask)
else:
    with torch.no_grad():
        x, out_len = self.model.wav2vec2.feature_extractor(waveforms, audio_lengths)
    padding_mask = components._get_padding_mask(x, out_len)
    x, attention_mask = self.model.wav2vec2.encoder._preprocess(x, out_len)
    x, mask = self.model.mask_generator(x, padding_mask)
    x = self.model.wav2vec2.encoder.transformer(x, attention_mask=attention_mask)
logits = self.aux(x)
```
Is there a better way to simplify the logic?

File: torchaudio/models/wav2vec2/model.py [hunk 6]
```diff
@@ -137,18 +137,27 @@ class HuBERTPretrainModel(Module):
 
         logit_generator (torch.nn.Module):
             Logit generator that predicts the logits of the masked and unmasked inputs.
+
+        feature_grad_mult (float or None):
```
Comment (hwangjeff) [inline torchaudio/models/wav2vec2/model.py:141]: can we explain the behavior when `feature_grad_mult` is `None`?

File: torchaudio/models/wav2vec2/model.py (hunk 4 above)
Comment (mthrok) [reply torchaudio/models/wav2vec2/model.py]: > Is there a better way to simplify the logic?

Since the data flow is same, I think one can set `requires_grad` outside of the loop when it hits the right condition.

The pseudo code would look like;
1. Initialize the `self.model.wav2vec2`
2. Disable gradient entirely
3. Run the training loop for `self.fix_encoder_updates`
4. Once the `self.fix_encoder_updates` step is done, enable gradient for `self.model.wav2vec2.feature_extractor`.

Comment (nateanl) [reply torchaudio/models/wav2vec2/model.py]: I see. I need to search where is the best place to enable gradient. `train_step` is called every iteration, so it's not suitable to call gradient enabling here.

to calrify, during the whole fine-tuning, `self.model.wav2vec2.feature_extractor` is freezed, otherwise the training will be unstable. In the first `fix_encoder_updates` iterations, the wav2vec2 model is freezed, only `aux` layer has gradient. after that, the gradient of `wav2vec2.encoder` is enabled.

**Comment by nateanl** [comment]
@mthrok do you want to merge https://github.com/pytorch/audio/pull/2153 first? then I adjust the annotations of hubert functions in this PR

**Review by mthrok — APPROVED** [review]
